packages/flutter_tools/lib/src/flutter_project_metadata.dart

```diff
@@ -140,17 +140,26 @@ class FlutterProjectMetadata {
 
   final Logger _logger;
 
-  final File _metadataFile;
+  final File? _metadataFile;
 
   /// Writes the .migrate_config file in the provided project directory's platform subdirectory.
   ///
   /// We write the file manually instead of with a template because this
   /// needs to be able to write the .migrate_config file into legacy apps.
   void writeFile({File? outputFile}) {
     outputFile = outputFile ?? _metadataFile;
+    if (outputFile == null) {
```

Comment:
Not entirely sure what's going on here, but this doesn't seem right. If this is an internal error that the user can't do anything about, then it should crash the tool by throwing an exception. If this is a problem the user can fix, but the tool can't continue, then this should be a `throwToolExit`.